x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1520,12 +1535,12 @@ public static Boolean isTokenServiceEnabled(Settings settings) {
                 clientAuthentication.getUser().principal()
             );
             return Optional.of(invalidGrantException("tokens must be refreshed by the creating client"));
-        } else if (clientAuthentication.getAuthenticatedBy().getName().equals(refreshToken.getAssociatedRealm()) == false) {
+        } else if (clientAuthentication.getSourceRealm().getName().equals(refreshToken.getAssociatedRealm()) == false) {
```

Comment:
I'm not sure if this is the change we want...
I think it was intentional that the client was checked for authentication not lookup.

Maybe that's wrong, but I'm not 100% sure it is a safe bug-fix to change that.

Reply:
We discussed in a separate channel and agreed to move this change ("client bound for refresh token") out of this PR ("client_credentials with run-as"). It will be addressed separately after further discussion.